src/query/boolean_query/boolean_weight.rs

```diff
@@ -112,13 +150,22 @@ impl<TScoreCombiner: ScoreCombiner> BooleanWeight<TScoreCombiner> {
     ) -> crate::Result<SpecializedScorer> {
         let mut per_occur_scorers = self.per_occur_scorers(reader, boost)?;
 
-        let should_scorer_opt: Option<SpecializedScorer> = per_occur_scorers
-            .remove(&Occur::Should)
-            .map(|scorers| scorer_union(scorers, &score_combiner_fn));
+        let should_scorer_opt: Option<SpecializedScorer> =
+            per_occur_scorers.remove(&Occur::Should).map(|scorers| {
+                if self.minimum_number_should_match <= 1 {
+                    scorer_union(scorers, &score_combiner_fn)
+                } else {
+                    SpecializedScorer::Other(scorer_disjunction(
+                        scorers,
+                        score_combiner_fn(),
+                        self.minimum_number_should_match,
+                    ))
+                }
+            });
```

Comment:
There are several missed optimization opportunities here:
if the number of scores is lower than `self.minimum_number_should_match`
we should early return an EmptyScorer.
If it is == then we can push process those scorers with the ones in the Must clause, as an intersection.